### PR TITLE
Visitor for authorization logic AST

### PR DESCRIPTION
**Pull request by aferr**

### CODE REVIEW

File: src/ir/ir_visitor.h
Comment (aferr) [inline src/ir/ir_visitor.h:34]: FYI This looked like a typo so I deliberately fixed this here.

Comment (Cypher1) [reply src/ir/ir_visitor.h:34]: Thanks! I wish cpp kept us honest about these.

**Review by Cypher1 — APPROVED** [review]
Lgtm

File: src/ir/auth_logic/auth_logic_ast_traversing_visitor.h
Comment (aferr) [reply src/ir/auth_logic/auth_logic_ast_traversing_visitor.h]: I started doing this, but because the body of the function you suggested calls `Accept` on element, the template really needs to specify that Element implements this method. One way to do this would be to write an abstract class that defines the methods all AST nodes have and include `Accept` in this abstract class and then have all the AST nodes extend from that. This seems like a more substantial change that belongs in another PR. So I wrote an issue to track this: #648

File: src/ir/auth_logic/auth_logic_ast_traversing_visitor.h
Comment (aferr) [reply src/ir/auth_logic/auth_logic_ast_traversing_visitor.h]: That doesn't quite work unfortunately... I tried what you suggested tweaking it slightly to the following because it's missing a param and a capture to the following:

```
    Result variant_visit_result = std::visit([*this](const auto &branch) {
      return VariantVisit(branch);
    }, base_fact.GetValue());
```

and auto is not able to deduce the type:
```
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1/variant:548:12: error: cannot deduce return type 'auto' from returned value of type '<overloaded function type>'
```

along with this other cryptic error I did not dig into:
```
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1/variant:659:14: error: attempt to use a deleted function
      return _VSTD::__invoke_constexpr(_VSTD::forward<_Visitor>(__visitor),
```

 I remember running into a similar issue when implementing `lowering_ast_datalog`.

I think actually the better way to remove this code would be to remove the use of `variant`s entirely from the implementation of the AST and replace them with sublcasses representing the different forms. This visitor is already implementing the pattern in a way that uses dynamic dispatch and inheritance to call the `Accept` implementation specific to the subclasses of the nodes. If you think that makes sense, I'll open an issue. Otherwise, it's left as-is.

File: src/ir/auth_logic/ast.h
Comment (aferr) [reply src/ir/auth_logic/ast.h]: fixed

File: src/ir/auth_logic/auth_logic_ast_traversing_visitor.h
Comment (aferr) [reply src/ir/auth_logic/auth_logic_ast_traversing_visitor.h]: fixed

File: src/ir/auth_logic/auth_logic_ast_traversing_visitor.h
Comment (aferr) [reply src/ir/auth_logic/auth_logic_ast_traversing_visitor.h]: see above. resolving this thread.

File: src/ir/auth_logic/auth_logic_ast_traversing_visitor.h
Comment (aferr) [reply src/ir/auth_logic/auth_logic_ast_traversing_visitor.h]: fixed

File: src/ir/auth_logic/auth_logic_ast_traversing_visitor.h
Comment (markww) [reply src/ir/auth_logic/auth_logic_ast_traversing_visitor.h]: Sounds good, if we're already using subclasses and vtables to subtype, might as well use them everywhere. I think that's a good way forward. Unfortunate that the generic lambda didn't work :(

File: src/ir/auth_logic/auth_logic_ast_traversing_visitor.h
Comment (markww) [reply src/ir/auth_logic/auth_logic_ast_traversing_visitor.h]: Templates don't need to specify anything; they'll just try subbing in the type implied by the arguments into the type parameter and see whether the code compiles.

Although, I do now realize that the order of the `Container` and `Element` type parameters should be swapped, as `Container` is inferrable from arguments and `Element` will probably need to be explicitly specified.

**Review by markww — APPROVED** [review]
LGTM modulo the last few (relatively minor) threads here.

File: src/ir/auth_logic/auth_logic_ast_traversing_visitor_test.cc
Comment (bgogul) [inline src/ir/auth_logic/auth_logic_ast_traversing_visitor_test.cc:32]: Redundant?

File: src/ir/auth_logic/auth_logic_ast_traversing_visitor_test.cc
Comment (bgogul) [inline src/ir/auth_logic/auth_logic_ast_traversing_visitor_test.cc:1]: I don't think the files are formatted. Can you use clang-format with google3 style to format them?

File: src/ir/auth_logic/auth_logic_ast_traversing_visitor_test.cc
Comment (bgogul) [inline src/ir/auth_logic/auth_logic_ast_traversing_visitor_test.cc:37]: Is it correct that these should be passed by value?

File: src/ir/auth_logic/auth_logic_ast_traversing_visitor_test.cc
Comment (bgogul) [inline src/ir/auth_logic/auth_logic_ast_traversing_visitor_test.cc:27]: Test is not comprehensive enough? It only tests previsit of a `Principal`?

**Review by bgogul — COMMENTED** [review]
Can you add some comprehensive tests? This is a lot of non-trivial code without sufficient tests.

File: src/ir/auth_logic/auth_logic_ast_traversing_visitor_test.cc
Comment (aferr) [reply src/ir/auth_logic/auth_logic_ast_traversing_visitor_test.cc:1]: should be OK now.

File: src/ir/auth_logic/auth_logic_ast_traversing_visitor_test.cc
Comment (aferr) [reply src/ir/auth_logic/auth_logic_ast_traversing_visitor_test.cc:29]: I made this more comprehensive.

File: src/ir/auth_logic/auth_logic_ast_traversing_visitor_test.cc
Comment (aferr) [reply src/ir/auth_logic/auth_logic_ast_traversing_visitor_test.cc:37]: I left this open in case I am missing something though, which might be the case.

File: src/ir/auth_logic/auth_logic_ast_traversing_visitor.h
Comment (aferr) [reply src/ir/auth_logic/auth_logic_ast_traversing_visitor.h]: Ah I mixed up templates with Generics from Java again. Swapping the order must have been it. It's implemented now.

**Comment by aferr** [comment]
@bgogul I'm not blocked on this yet so I thought I would wait for you to re-review before merging it in. Let me know if it looks OK if you have a moment.

File: src/ir/auth_logic/auth_logic_ast_traversing_visitor_test.cc
Comment (bgogul) [inline src/ir/auth_logic/auth_logic_ast_traversing_visitor_test.cc:187]: It might be more straightforward with a gmock, but it is fine for now. 

The only question is if there is a way to break this down into smaller tests.

**Review by bgogul — APPROVED** [review]
Approving, but left some clean up comments. If you don't address these, can you file issues (e.g., gmock, debugprint as visitor, etc.)

Also, try to keep PRs smaller.

File: src/ir/auth_logic/ast.h
Comment (aferr) [reply src/ir/auth_logic/ast.h:191]: Right, this is just for testing the visitor for now.

File: src/ir/auth_logic/auth_logic_ast_traversing_visitor.h
Comment (aferr) [reply src/ir/auth_logic/auth_logic_ast_traversing_visitor.h:19]: fixed

File: src/ir/auth_logic/auth_logic_ast_traversing_visitor_test.cc
Comment (aferr) [reply src/ir/auth_logic/auth_logic_ast_traversing_visitor_test.cc:29]: I'm assuming it's enough and resolving :)

File: src/ir/auth_logic/auth_logic_ast_traversing_visitor_test.cc
Comment (aferr) [reply src/ir/auth_logic/auth_logic_ast_traversing_visitor_test.cc:37]: I'm assuming this was OK and resolving.

File: src/ir/auth_logic/auth_logic_ast_traversing_visitor_test.cc
Comment (aferr) [reply src/ir/auth_logic/auth_logic_ast_traversing_visitor_test.cc:187]: Opened an issue just because there is another PR depending on this one (and IDK or don't remember what a gmock is yet) #666

**Comment by aferr** [comment]
> Also, try to keep PRs smaller.

@bgogul Writing code in smaller increments doesn't come so naturally to me. This one seems like a self-contained increment to me. How would I split this into smaller self-contained PRs? Would it be to separately write:
 - just the visitor
 - then the traversing visitor + changes to the AST nodes
 - then add the test
Or what would be a sensible way to split this into smaller PRs?

**Pull request merged**
